Plugins~/Src/MeshSyncClientBlender/msblenMaterialsExportHelper.cpp

```diff
@@ -0,0 +1,497 @@
+
+			if (name.find(extension) != name.length() - extension.length()) {
+				name += extension;
+			}
+		}
+
+		int exported = m_texture_manager->addImage(name, 0, 0, imagePackedFile->packedfile->data, imagePackedFile->packedfile->size, ms::TextureFormat::RawFile, textureType);
+		setTextureHandler(exported);
+	}
+}
+
+void msblenMaterialsExportHelper::exportImageFromImageNode(ms::TextureType& textureType,
+	std::function<void(int textureId)> setTextureHandler,
+	bNode* sourceNode) const
+{
+	if (setTextureHandler) {
+		auto img = (Image*)sourceNode->id;
+
+		// Use non-color if the image node is not in sRGB space:
+		if (textureType == ms::TextureType::Default && !STREQ(img->colorspace_settings.name, "sRGB")) {
+			textureType = ms::TextureType::NonColor;
+		}
+
+		// Unpack if needed:
+		if (img->packedfiles.first) {
+			exportPackedImages(textureType, setTextureHandler, img);
+		}
+		else {
+			setTextureHandler(exportTexture(abspath(img->filepath), textureType));
+		}
+	}
+}
+
+void msblenMaterialsExportHelper::handleImageNodeWithAssignedImage(ms::TextureType& textureType,
+	bool resetIfInputIsTexture,
+	std::function<void(const mu::float4& colorValue)> setColorHandler,
+	std::function<void(int textureId)> setTextureHandler,
+	bNode* sourceNode) const
+{
+	exportImageFromImageNode(textureType, setTextureHandler, sourceNode);
+
+	// Ensure the color is white if there is a texture otherwise Unity will multiply that image with the color that was set before.
+	// Blender does not do that so they would not look the same:
+	if (resetIfInputIsTexture && setColorHandler) {
+		setColorHandler(mu::float4{ 1, 1, 1, 1 });
+	}
+}
+
+void msblenMaterialsExportHelper::handleImageNode(ms::TextureType& textureType,
+	bool resetIfInputIsTexture,
+	std::function<void(const mu::float4& colorValue)> setColorHandler,
+	std::function<void(int textureId)> setTextureHandler,
+	bNode* sourceNode)
+{
+	if (sourceNode->id) {
+		handleImageNodeWithAssignedImage(textureType, resetIfInputIsTexture, setColorHandler, setTextureHandler, sourceNode);
+	}
+	else {
+		if (setTextureHandler) {
+			setTextureHandler(ms::InvalidID);
+		}
+
+		if (setColorHandler) {
+			// Blender uses black if there is no image set on an image node:
+			setColorHandler(mu::float4{ 0, 0, 0, 1 });
+		}
+	}
+}
+
+void msblenMaterialsExportHelper::handleNormalMapNode(const Material* mat,
+	ms::TextureType textureType,
+	bool resetIfInputIsTexture,
+	std::function<void(const mu::float4& colorValue)> setColorHandler,
+	std::function<void(int textureId)> setTextureHandler,
+	bNode* sourceNode)
+{
+	if (setTextureHandler) {
+		auto imageInput = getInputSocket(sourceNode, colorIdentifier);
+		if (imageInput)
+		{
+			setValueFromSocket(mat,
+				imageInput, textureType,
+				resetIfInputIsTexture,
+				nullptr,
+				setTextureHandler);
+		}
+	}
+
+	auto strengthInput = getInputSocket(sourceNode, normalStrengthIdentifier);
+	if (strengthInput && setColorHandler) {
```

Comment:
Can you do this same sort of logic on the if statement above to be consistent & reduce nesting?

Reply:
Do it where? Not sure I understand.

Reply:
```
auto imageInput = getInputSocket(sourceNode, colorIdentifier);
if (setTextureHandler && imageInput) {
        setValueFromSocket(mat,
				imageInput, textureType,
				resetIfInputIsTexture,
				nullptr,
				setTextureHandler);	
}
```

Reply:
Ah ok, done